# schickit🚀

---

参考脚本和结果: /cluster/home/tmp/ckw/tutorial_result

参考数据: /cluster/home/tmp/ckw/tutorial

参考基因存放位置: /cluster/home/tmp/ckw/common

# Table of contents
* [Introduction](#introduction)
* [Citation](#citation)
* [Installation](#installation)
* [Usage](#usage)
    * [Prepare index and enzyme bed](#prepare-index-and-enzyme-bed)
    * [Typical Workflow](#typical-workflow)
        * [1. scHic_no 🪝](#1-schicno-)
        * [2. scHic Index 🪝](#2-schic-with-index-)
        * [3. dipC 🪝](#3-dipc-)
        * [4. HIRES 🪝](#4-hires-)
        * [5. sn_m3c 🪝](#5-snm3c-)
        * [6. scSPRITE 🪝](#6-scsprite-)
        * [7. sciHiC 🪝](#7-scihic-)
        * [8. snHiC_no 🪝](#8-snhicno-)
        * [9. snHiC_index 🪝](#9-snhicindex-)
        * [10. scNano 🪝](#10-scnano-)
    * [Result](#result)
# Introduction

--- 

Schickit is a  software for processing various types of single-cell Hi-C sequencing data, which currently includes the  single-cell Hi-C technologies as follows:

- scHic_no
- scHic_index
- dipC
- HIRES
- sn_m3c
- scSPRITE
- sciHic
- snHic_no
- snHic_index
- scNano

At the same time, we also provide a generalized way to process the single-cell Hi-C, for details please see the [tutorial](#typical-workflow).

To get started：

- Install schickit
- Read the documentation and see the Jupyter Notebook walkthrough.
- test data can be download form [Zenodo](https://zenodo.org/records/10426561/files/tutorial.tar.gz)
- Many more single-cell Hi-C data are available on our website

# Installation

---

We suggest creating a new python environment to install schickit before installation and schickit is based on python≥3.9

```shell
python create -n schickit python>=3.9
conda activate schickit
```

schickit is based on software such as bwa, bowtie2, minimap2, fastp, hicexploer, samtools, bedtools, bedtools, bismark, nanoplexer, etc. Please make sure you have installed the corresponding software before using it. You can write the following code into a shell script and run it.

```shell
# install bwa
yes|conda install -c bioconda bwa
# install bowtie2
yes|conda install -c bioconda bowtie2
# install minimap2
yes|conda install -c bioconda minimap2
# install fastp
yes|conda install -c bioconda fastp
# install hicexplorer
yes|conda install -c bioconda hicexplorer
# install samtools
yes|conda install -c bioconda samtools
# install bedtools
yes|conda install -c bioconda bedtools
# install bismark
yes|conda install -c bioconda bismark
# install nanoplexer
yes|conda install -c bioconda nanoplexer
```

Then install schickit

```shell
    pip install schickit
```

# Citation

---

Not written yet.
The info of test data


# Usage

---

### Prepare index and enzyme bed

schikit has its own format for reading reference genomes and enzyme bed files. For convenience, we provide an automated program for building indexes and enzyme bed files:

```shell
schickit index -g hg38 -a bwa -p ./save/path -e MboI,DpnII,BglII,HpyCH4V
```

Parameter description:

- -g: The genome version you want to assemble, currently supports hg38, hg19, mm10, mm9.
- -a: The software used for assembly, different software produces different index files
- -p: Path where the results are stored
- -e: The required enzyme bed file, can be single or multiple inputs (different genomes correspond to different enzyme files)

## **Typical Workflow**

---

### 1. scHic_no 🪝

Test data: /tutorial/scHic_no/data

```bash
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t scHic \
	-e MboI \
	-r 10000 \
	-i /path/to/index/hg38.fa \
	-a bwa \
	--thread 60
```

Parameter Description:

- -o Location to save the result, note that **all paths must be absolute paths.**
- -f Directory where the sequencing data is located
- -t Type of single-cell Hi-C
- -e Type of restriction enzyme used, must be consistent with the experiment
- -r Resolution used to convert pairs files into cool files
- -i Directory of the genome file used for alignment, the final hg38.fa is the type of genome, not part of the directory, consistent with the -g parameter of schickit index
- -a The software used for assembly, optional bwa, bowtie2, bismark, minimap2. Here it should be consistent with the index produced by schickit index

### 2. scHic with index 🪝

Test data: /tutorial/scHic_index/data

It should be noted that the number of fastq files generated by this technique should be 4, where _1 and _4 are reads, and _2 and _3 are the corresponding barcode files.

```shell
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t scHic \
	-e MboI \
	-r 10000 \
	-i /path/to/index/hg38.fa \
	--aligner bwa \
	--thread 60 \
	--exist-barcode
```

Note that:

The only difference with scHic_no is that you need to specify that it has barcodes wtih parameter —exist-barcode.

### 3. dipC 🪝

Test data: /tutorial/dipC/data

```shell
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t dipC \
	-e MboI \
	-r 10000 \
	-i /path/to/index/hg38.fa \
	--aligner bwa \
	--thread 60
```

Similar with scHic_no, schickit will automatically process the following steps according to the parameter -t.

### 4. HIRES 🪝

Test data: /tutorial/HRIES/data

```shell
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t HIRES \
	-e MboI \
	-r 10000 \
	-i /path/to/index/mm10.fa \
	-a bwa \
	--thread 60
```

Similar with scHic_no, schickit will automatically process the follow-up according to the parameter -t.

### 5. sn_m3c 🪝

Test data: /tutorial/sn_m3c/data

```shell
schickit count -o /path/to/save \
			-f /path/to/fastq \
			-t sn_m3c \
			-e MboI \
			-r 10000 \
			-i /path/to/index/hg38.fa \
			--aligner bowtie2 \
			--thread 60
```

Since the sn_m3c sequencing simultaneously methylation and Hi-C, only bismark can be used for assembly. Correspondingly, there is no multiple choice for -i and -a. However, bismark is based on bowtie2, you can write the parameters as bowtie2.

### 6. scSPRITE 🪝

Test data: /tutorial/scSPRITE/data

```shell
schickit count -o /path/to/save \
        -f /path/to/fastq \
		-t scSPRITE \
		-e HpyCH4V \
		-r 10000 \
		-i /path/to/index/mm10.fa \
		-a bwa \
		--thread 60 \
		--sprite-config /path/to/config/config_dpm6_y-stag_scSPRITE2.txt \
		--repeat-masker /path/to/repeat_mask/mm10_rmsk.bed
```

Parameter Description:
- --sprite-config is a txt file for scSPRITE generating barcode.
- --repeat-masker is a bed file for genome masking of repetitive regions
### 7. sciHic 🪝

Test data: /tutorial/sciHic/data. It should be noted that not only two sequencing fastq files are needed in each sequencing file directory, but also txt files corresponding to inner_barcode.txt and outer_barcode.txt are needed. The format is as shown in the example.

```shell
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t sciHic \
	-e DpnII \
	-r 10000 \
	-i /path/to/index/mm10.fa \
	--aligner bwa \
	--thread 60 \
	--exist-barcode
```

Similar with scHic_index, schickit will automatically carry out subsequent processing according to the parameter -t.

### 8. snHic_no 🪝

Test data: /tutorial/snHic_no/data

```shell
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t snHic \
	-e DpnII \
	-r 10000 \
	-i /path/to/index/hg38.fa \
	--aligner bwa \
	--thread 60
```

Similar with scHic_no, schickit will automatically process the follow-up according to the parameter -t.

### 9. snHic_index 🪝

Test data: /tutorial/snHic_index/data

```shell
schickit count -o /path/to/save \
	-f /path/to/fastq \
	-t snHic \
	-e MboI \
	-r 10000 \
	-i /path/to/index/hg38.fa \
	--aligner bwa \
	--thread 60 \
	--exist-barcode
```

Similar with scHic_no, schickit will automatically process the following steps according to the parameter -t.

### 10. scNano 🪝

Test data: /tutorial/scNano/data

It is worth noting that scNano uses third-generation sequencing, and the sequnencing data directory not only should contain the fastq file, but alse the TN5.txt, PCR.txt and index.txt

```shell
schickit count -o /path/to/save \
    -f /path/to/fastq \
	-t scNano \
	-e MboI \
	-r 10000 \
	-i /path/to/index/mm10.fa \
	-a bwa \
	--thread 60 
```

scNano uses third-generation sequencing technology, and uses minimap2 by default. So  the -a parameter will become ineffective.

## Result

---

Here is the introduction to the results:

``` shell
scSPRITE_test_tmp/
├── Result
│   ├── cool_folder
│       ├── [Even2Bo10][Odd2Bo69][DPM6bot1]_10000.cool
│       ├── [Even2Bo10][Odd2Bo69][DPM6bot1]10000.KR.cool
│       ├── [Even2Bo11][Odd2Bo19][DPM6bot31]_10000.cool
│       ├── [Even2Bo11][Odd2Bo19][DPM6bot31]10000.KR.cool
│       ├── [Even2Bo11][Odd2Bo1][DPM6bot75]_10000.cool
│       ├── [Even2Bo11][Odd2Bo1][DPM6bot75]10000.KR.cool
│   ├── mcool_folder
│   │   ├── [Even2Bo10][Odd2Bo69][DPM6bot1].mcool
│   │   ├── [Even2Bo11][Odd2Bo19][DPM6bot31].mcool
│   │   ├── [Even2Bo11][Odd2Bo1][DPM6bot75].mcool
│   └── SCpair
│       ├── [Even2Bo10][Odd2Bo69][DPM6bot1].pairs.gz
│       ├── [Even2Bo11][Odd2Bo19][DPM6bot31].pairs.gz
│       ├── [Even2Bo11][Odd2Bo1][DPM6bot75].pairs.gz
├── test.bam
├── test_logging.log
└── trimmed.fastp.json
```

***scSPRITE_test_tmp*** is the root directory of the output, where ‘test’ is the name of the sample.

***Result*** is directory the main result saved in it.

***cool_folder*** is the directory stores all cells’ cool files before and after **KR** correction.

***mcool _folder*** is the directory stores all cells’ mcool files.

***SCpair*** is the directory stores all cells’ pair files.

***test.bam*** is the bam file of the sequencing data.

***test_logging.log*** records all the parameters of processing the data by schickit, as well as the time spent on each step.

***trimmed.fastp.json*** records the result of fastp processing the data.